tests/common/util.rs

```diff
@@ -1183,72 +1201,90 @@ pub struct UCommand {
-                        }
-                    }
-                }
-                cmd
-            },
-            comm_string: String::from(bin_path.to_str().unwrap()),
-            bin_path: bin_path.to_str().unwrap().to_string(),
-            util_name: util_name.map(|un| un.to_str().unwrap().to_string()),
-            ignore_stdin_write_error: false,
-            bytes_into_stdin: None,
-            stdin: None,
-            stdout: None,
-            stderr: None,
-            #[cfg(any(target_os = "linux", target_os = "android"))]
-            limits: vec![],
-            stderr_to_stdout: false,
-            timeout: Some(Duration::from_secs(30)),
-        };
-
-        if let Some(un) = util_name {
-            ucmd.arg(un);
+    /// Create a new plain [`UCommand`].
+    ///
+    /// If not specified otherwise with [`UCommand::bin_path`] or [`UCommand::util`] the underlying
+    /// [`Command] executes a command that must be given as argument (for example with
+    /// [`UCommand::arg`] in a shell (`sh -c` on unix platforms or `cmd /C` on windows).
+    ///
+    /// Per default the environment is cleared and the working directory is set to an individual
+    /// temporary directory for safety purposes.
+    pub fn new() -> Self {
+        Self {
+            env_clear: true,
+            ..Default::default()
         }
+    }
 
+    /// Create a [`UCommand`] for a specific uutils utility.
+    ///
+    /// Sets the temporary directory to `tmpd` and the execution binary to the path where
+    /// `coreutils` is found.
+    pub fn with_util<T>(util_name: T, tmpd: Rc<TempDir>) -> Self
+    where
+        T: AsRef<str>,
+    {
+        let mut ucmd = Self::new();
+        ucmd.util(util_name).temp_dir(tmpd);
         ucmd
     }
 
-    pub fn new_from_tmp<T: AsRef<OsStr>, S: AsRef<OsStr>>(
-        bin_path: T,
-        util_name: &Option<S>,
-        tmpd: Rc<TempDir>,
-        env_clear: bool,
-    ) -> Self {
-        let tmpd_path_buf = String::from(tmpd.as_ref().path().to_str().unwrap());
-        let mut ucmd: Self = Self::new(bin_path, util_name, tmpd_path_buf, env_clear);
-        ucmd.tmpd = Some(tmpd);
-        ucmd
+    /// Create a [`UCommand`] from a [`TestScenario`].
+    ///
+    /// The temporary directory and uutils utility are inherited from the [`TestScenario`] and the
+    /// execution binary is set to `coreutils`.
+    pub fn from_test_scenario(scene: &TestScenario) -> Self {
+        Self::with_util(&scene.util_name, scene.tmpd.clone())
+    }
+
+    /// Set the execution binary.
+    ///
+    /// Make sure the binary found at this path is executable. It's safest to provide the
+    /// canonicalized path instead of just the name of the executable, since path resolution is not
+    /// guaranteed to work on all platforms.
+    pub fn bin_path<T>(&mut self, bin_path: T) -> &mut Self
+    where
+        T: Into<PathBuf>,
+    {
+        self.bin_path = Some(bin_path.into());
+        self
+    }
```

Comment:
Are these setters necessary (also `util`, `temp_dir`)? Or can we maybe not make them `pub`?

Reply:
`util` could be private, `with_util` and `from_test_scenario` should be sufficient. As far as I can see, `temp_dir` could be private too if you think it's not useful to be `pub`. Same for `bin_path` (`TestScenario::cmd` provides the same functionality but then the only way to set a custom binary would be via `TestScenario`). `current_dir` is actually useful, since it removes the need for access to the raw `std::process::Command`. Compared to the public api of `Command` this is the last missing method and it's in use in `test_cp` and `test_pwd`.

We could reduce the public api of `TestScenario` too. `ucmd_keepenv` and `cmd_keepenv` are much less useful, since there's `keep_env` in `UCommand`, now.

Reply:
That sounds great!